### DISCUSSION TITLE
SIMD or NDBuffer which one is more like ndarray in numpy?

### DISCUSSION THREAD

**Discussion by yt7589** [modular/modular#236] [Q&A]
I wan to have the functionality of ndarray in numpy. I found that there has no such things in mojo. If I want to develop a NDArray in mojo which one is prefered: SIMD or NDBuffer? According to my knowledge: SIMD is 1D array but has operation (such as +, *, broadcast) override function. NDBuffer is nD array but lack operation override function. Please give me some advice.

**Comment by mojodojodev** [comment]
There are no high level libraries like Numpy in Mojo yet, the closest is NDBuffer but it's very low level and doesn't own it's data. For some examples on how to use it check this: https://mojodojo.dev/guides/std/Buffer/NDBuffer.html

**Reply by guidorice** [reply]
Updated link: https://mojodojo.dev/guides/std/Buffer/NDBuffer.html

I am curious how the `Tensor` in today's release compares with your tensor example. And indeed, compares with the Matrix example in the Matrix multiplication notebook.

**Reply by guidorice** [reply]
This was summarized in https://mojodojo.dev/this-week-in-mojo/2023-08-11.html as well. 👍🏽

**Reply by mojodojodev** [reply]
Thanks @guidorice updated the link, I’ll do a guide using Tensor, it’s similar
